### PR TITLE
Tune Arithmetic Op launch specification

### PR DESCRIPTION
**Pull request by klecki**
Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

#### Why we need this PR?
Arithmetic Op is not fast enough, GPU was underutilized due to not big enough tiles.

#### What happened in this PR?
 - What solution was applied:
     Tile was made bigger, grid was made smaller to allow for few more iterations for given thread.
     Pointer usage was adjusted a bit, it helps with small inputs a bit.
     Benchmark was added
 - Affected modules and functionalities:
     Arithmetic Ops
 - Key points relevant for the review:
     
 - Validation and testing:
     Benchmark
 - Documentation (including examples):
     NA

OLD:
```
test_operator_arithmetic_ops.test_arithmetic_ops_perf(('gpu', 'const'), (<class 'numpy.float32'>, <class 'numpy.float32'>), <function test_arithmetic_ops_perf.<locals>.<lambda> at 0x7f5fd44f6ae8>, [(1024, 1024)] * 256, '*') ... Throughput: 377.034 GB/s
Throughput: 375.006 GB/s
ok
test_operator_arithmetic_ops.test_arithmetic_ops_perf(('gpu', 'const'), (<class 'numpy.float32'>, <class 'numpy.float32'>), <function test_arithmetic_ops_perf.<locals>.<lambda> at 0x7f5fd44f6ae8>, [(16384, 1024)] * 64, '*') ... Throughput: 391.476 GB/s
Throughput: 405.331 GB/s
ok
test_operator_arithmetic_ops.test_arithmetic_ops_perf(('gpu', 'const'), (<class 'numpy.float32'>, <class 'numpy.float32'>), <function test_arithmetic_ops_perf.<locals>.<lambda> at 0x7f5fd44f6ae8>, [(400, 400)] * 64, '*') ... Throughput: 334.448 GB/s
Throughput: 353.607 GB/s
ok
test_operator_arithmetic_ops.test_arithmetic_ops_perf(('gpu', 'gpu'), (<class 'numpy.float32'>, <class 'numpy.float32'>), <function test_arithmetic_ops_perf.<locals>.<lambda> at 0x7f5fd3e962f0>, [(1024, 1024)] * 256, '*') ... Throughput: 490.778 GB/s
Throughput: 490.198 GB/s
ok
test_operator_arithmetic_ops.test_arithmetic_ops_perf(('gpu', 'gpu'), (<class 'numpy.float32'>, <class 'numpy.float32'>), <function test_arithmetic_ops_perf.<locals>.<lambda> at 0x7f5fd3e962f0>, [(16384, 1024)] * 64, '*') ... Throughput: 499.623 GB/s
Throughput: 507.385 GB/s
ok
test_operator_arithmetic_ops.test_arithmetic_ops_perf(('gpu', 'gpu'), (<class 'numpy.float32'>, <class 'numpy.float32'>), <function test_arithmetic_ops_perf.<locals>.<lambda> at 0x7f5fd3e962f0>, [(400, 400)] * 64, '*') ... Throughput: 358.166 GB/s
Throughput: 402.577 GB/s
ok

----------------------------------------------------------------------
Ran 6 tests in 10.587s

OK
```
NEW:
```
test_operator_arithmetic_ops.test_arithmetic_ops_perf(('gpu', 'const'), (<class 'numpy.float32'>, <class 'numpy.float32'>), <function test_arithmetic_ops_perf.<locals>.<lambda> at 0x7f96c80bbae8>, [(1024, 1024)] * 256, '*') ... Throughput: 533.855 GB/s
Throughput: 531.948 GB/s
ok
test_operator_arithmetic_ops.test_arithmetic_ops_perf(('gpu', 'const'), (<class 'numpy.float32'>, <class 'numpy.float32'>), <function test_arithmetic_ops_perf.<locals>.<lambda> at 0x7f96c80bbae8>, [(16384, 1024)] * 64 '*') ... Throughput: 576.293 GB/s
Throughput: 576.419 GB/s
ok
test_operator_arithmetic_ops.test_arithmetic_ops_perf(('gpu', 'const'), (<class 'numpy.float32'>, <class 'numpy.float32'>), <function test_arithmetic_ops_perf.<locals>.<lambda> at 0x7f96c80bbae8>, [(400, 400)] * 64, '*') ... Throughput: 462.535 GB/s
Throughput: 498.008 GB/s
ok
test_operator_arithmetic_ops.test_arithmetic_ops_perf(('gpu', 'gpu'), (<class 'numpy.float32'>, <class 'numpy.float32'>), <function test_arithmetic_ops_perf.<locals>.<lambda> at 0x7f96c7a5c2f0>, [(1024, 1024)] * 256, '*') ... Throughput: 571.668 GB/s
Throughput: 572.734 GB/s
ok
test_operator_arithmetic_ops.test_arithmetic_ops_perf(('gpu', 'gpu'), (<class 'numpy.float32'>, <class 'numpy.float32'>), <function test_arithmetic_ops_perf.<locals>.<lambda> at 0x7f96c7a5c2f0>, [(16384, 1024)] * 64, '*') ... Throughput: 604.232 GB/s
Throughput: 603.718 GB/s
ok
test_operator_arithmetic_ops.test_arithmetic_ops_perf(('gpu', 'gpu'), (<class 'numpy.float32'>, <class 'numpy.float32'>), <function test_arithmetic_ops_perf.<locals>.<lambda> at 0x7f96c7a5c2f0>, [(400, 400)] * 64, '*') ... Throughput: 447.778 GB/s
Throughput: 459.841 GB/s
ok

----------------------------------------------------------------------
Ran 6 tests in 10.426s

OK
```

**JIRA TASK**: *[Use DALI-1514 or NA]*

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: dali/operators/math/expressions/expression_impl_gpu.cuh [hunk 1]
```diff
@@ -66,16 +80,20 @@ __device__ void ExecuteBinOp(Result *result, Left l, const Right *r, int64_t ext
 template <ArithmeticOp op, typename Result, typename Left, typename Right>
 __device__ void ExecuteBinOp(Result *result, const Left *l, Right r, int64_t extent) {
   using meta = arithm_meta<op, GPUBackend>;
+  result += blockIdx.x * blockDim.x + threadIdx.x;
+  l += blockIdx.x * blockDim.x + threadIdx.x;
   for (int64_t i = blockIdx.x * blockDim.x + threadIdx.x; i < extent; i += blockDim.x * gridDim.x) {
-    result[i] = meta::impl(l[i], r);
+    *result = meta::impl(*l, r);
```
Comment (mzient) [inline dali/operators/math/expressions/expression_impl_gpu.cuh:86]: Why? Does it help in any way?

Comment (klecki) [reply dali/operators/math/expressions/expression_impl_gpu.cuh:86]: It seems so with smaller inputs, with bigger ones it doesn't make much difference.

File: dali/operators/math/expressions/expression_impl_gpu.cuh [hunk 2]
```diff
@@ -33,8 +33,12 @@ namespace dali {
 template <ArithmeticOp op, typename Result, typename Input>
 __device__ void ExecuteUnOp(Result *result, const Input *in, int64_t extent) {
   using meta = arithm_meta<op, GPUBackend>;
+  result += blockIdx.x * blockDim.x + threadIdx.x;
+  in += blockIdx.x * blockDim.x + threadIdx.x;
   for (int64_t i = blockIdx.x * blockDim.x + threadIdx.x; i < extent; i += blockDim.x * gridDim.x) {
```
Comment (jantonguirao) [inline dali/operators/math/expressions/expression_impl_gpu.cuh]: extract `blockDim.x * gridDim.x` to a variable?

File: dali/operators/math/expressions/expression_impl_gpu.cuh [hunk 3]
```diff
@@ -33,8 +33,12 @@ namespace dali {
...
-    result[i] = meta::impl(in[i]);
```
Comment (jantonguirao) [inline dali/operators/math/expressions/expression_impl_gpu.cuh:37]: actually it seemed more readable before. Why the change?

File: dali/operators/math/expressions/expression_impl_gpu.cuh [hunk 4]
```diff
@@ -44,8 +48,14 @@ __device__ void ExecuteUnOp(Result *result, const Input *in, int64_t extent) {
 template <ArithmeticOp op, typename Result, typename Left, typename Right>
 __device__ void ExecuteBinOp(Result *result, const Left *l, const Right *r, int64_t extent) {
   using meta = arithm_meta<op, GPUBackend>;
+  result += blockIdx.x * blockDim.x + threadIdx.x;
+  l += blockIdx.x * blockDim.x + threadIdx.x;
+  r += blockIdx.x * blockDim.x + threadIdx.x;
   for (int64_t i = blockIdx.x * blockDim.x + threadIdx.x; i < extent; i += blockDim.x * gridDim.x) {
-    result[i] = meta::impl(l[i], r[i]);
+    *result = meta::impl(*l, *r);
```
Comment (jantonguirao) [inline dali/operators/math/expressions/expression_impl_gpu.cuh:61]: same applies here. Why the change?

File: dali/operators/math/expressions/arithmetic_test.cu [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    Fill(left.cpu(), fill_left);
+    Fill(right.cpu(), fill_right);
+
+    // Fill pointers for tiles
+    tiles_data.reshape(uniform_list_shape<1>(1, {TestConfig::num_tiles}));
+    auto tiles_cpu = tiles_data.cpu()[0];
+    for (int sample_id = 0; sample_id < TestConfig::batch_size; sample_id++) {
+      for (int extent_id = 0; extent_id < TestConfig::tiles_per_sample; extent_id++) {
+        int tile_id = sample_id * TestConfig::tiles_per_sample + extent_id;
+        tiles_cpu(tile_id)->desc = tile_descs[tile_id];
```
Comment (JanuszL) [inline dali/operators/math/expressions/arithmetic_test.cu:98]: Can't you fill `tile_descs` here as well?

File: dali/operators/math/expressions/expression_impl_gpu.cuh [hunk 6]
```diff
@@ -106,30 +124,28 @@ struct argument<true> {
...
 template <ArithmeticOp op, typename Result, typename Left, typename Right, bool IsLeftTensor,
           bool IsRightTensor>
-__global__ void ExecuteTiledBinOp(const ExtendedTileDesc *tiles, int num_tiles) {
+__global__ void ExecuteTiledBinOp(const ExtendedTileDesc *tiles) {
   const auto &tile = tiles[blockIdx.y];
-  auto output = static_cast<Result *>(tile.output);
-  auto left = static_cast<const Left *>(tile.args[0]);
-  auto right = static_cast<const Right *>(tile.args[1]);
+  auto *output = static_cast<Result *>(tile.output);
+  const auto *left = static_cast<const Left *>(tile.args[0]);
```
Comment (jantonguirao) [inline dali/operators/math/expressions/expression_impl_gpu.cuh:142]: I'd go with either
```suggestion
  const Left *left = static_cast<const Left *>(tile.args[0]);
```
or 
```suggestion
  auto left = static_cast<const Left *>(tile.args[0]);
```

File: dali/operators/math/expressions/expression_impl_gpu.cuh (hunk 1 above)
Comment (jantonguirao) [reply dali/operators/math/expressions/expression_impl_gpu.cuh:98]: In that case, can we store the offset and step in variables to avoid repeating?

Comment (klecki) [reply dali/operators/math/expressions/expression_impl_gpu.cuh:98]: Done I guess.

File: dali/operators/math/expressions/expression_impl_gpu.cuh (hunk 3 above)
Comment (klecki) [reply dali/operators/math/expressions/expression_impl_gpu.cuh:37]: For a small perf gain as well as correctness with int64 offset.

File: dali/operators/math/expressions/expression_impl_gpu.cuh (hunk 6 above)
Comment (klecki) [reply dali/operators/math/expressions/expression_impl_gpu.cuh:142]: I can go back to `auto left`

File: dali/operators/math/expressions/arithmetic_test.cu (hunk 5 above)
Comment (klecki) [reply dali/operators/math/expressions/arithmetic_test.cu:98]: I could but implementing that lambda would be probably a bit more of a hassle than a direct loop.

File: dali/operators/math/expressions/expression_impl_gpu.cuh [hunk 7]
```diff
@@ -44,8 +51,17 @@ __device__ void ExecuteUnOp(Result *result, const Input *in, int64_t extent) {
 template <ArithmeticOp op, typename Result, typename Left, typename Right>
 __device__ void ExecuteBinOp(Result *result, const Left *l, const Right *r, int64_t extent) {
   using meta = arithm_meta<op, GPUBackend>;
-  for (int64_t i = blockIdx.x * blockDim.x + threadIdx.x; i < extent; i += blockDim.x * gridDim.x) {
-    result[i] = meta::impl(l[i], r[i]);
+  uint32_t start_ofs = (blockDim.x) * blockIdx.x + threadIdx.x;
```
Comment (jantonguirao) [inline dali/operators/math/expressions/expression_impl_gpu.cuh:54]: in the previous implementation, you used int64_t. Can we use int64_t everywhere?

File: dali/operators/math/expressions/expression_impl_gpu.cuh [hunk 8]
```diff
@@ -55,8 +71,15 @@ __device__ void ExecuteBinOp(Result *result, const Left *l, const Right *r, int6
 template <ArithmeticOp op, typename Result, typename Left, typename Right>
 __device__ void ExecuteBinOp(Result *result, Left l, const Right *r, int64_t extent) {
   using meta = arithm_meta<op, GPUBackend>;
-  for (int64_t i = blockIdx.x * blockDim.x + threadIdx.x; i < extent; i += blockDim.x * gridDim.x) {
-    result[i] = meta::impl(l, r[i]);
+  uint32_t start_ofs = (blockDim.x) * blockIdx.x + threadIdx.x;
```
Comment (jantonguirao) [inline dali/operators/math/expressions/expression_impl_gpu.cuh:74]: same here

File: dali/operators/math/expressions/expression_impl_gpu.cuh [hunk 9]
```diff
@@ -66,16 +89,23 @@ __device__ void ExecuteBinOp(Result *result, Left l, const Right *r, int64_t ext
 template <ArithmeticOp op, typename Result, typename Left, typename Right>
 __device__ void ExecuteBinOp(Result *result, const Left *l, Right r, int64_t extent) {
   using meta = arithm_meta<op, GPUBackend>;
-  for (int64_t i = blockIdx.x * blockDim.x + threadIdx.x; i < extent; i += blockDim.x * gridDim.x) {
-    result[i] = meta::impl(l[i], r);
+  uint32_t start_ofs = (blockDim.x) * blockIdx.x + threadIdx.x;
```
Comment (jantonguirao) [inline dali/operators/math/expressions/expression_impl_gpu.cuh:92]: and here

File: dali/operators/math/expressions/expression_impl_gpu.cuh (hunk 7 above)
Comment (klecki) [reply dali/operators/math/expressions/expression_impl_gpu.cuh:54]: Added by mistake, will revert to previous int64_t change.

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

File: dali/operators/math/expressions/arithmetic_test.cu [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+    Right r{};
+    auto fill_left = [&l]() { return l += 1; };
+    auto fill_right = [&r]() { return r += 1; };
+    Fill(left.cpu(), fill_left);
+    Fill(right.cpu(), fill_right);
+
+    // Fill pointers for tiles
+    tiles_data.reshape(uniform_list_shape<1>(1, {TestConfig::num_tiles}));
+    auto tiles_cpu = tiles_data.cpu()[0];
+    // TestTensorList just allocates memory, this can leave SmallVector in weird state
```
Comment (klecki) [inline dali/operators/math/expressions/arithmetic_test.cu:95]: Still need a resize or push_back()

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build
